upstairs/src/lib.rs

```diff
@@ -1652,15 +1652,23 @@ impl Buffer {
         assert_eq!(self.block_size, buffer.block_size);
 
         let start_block = offset / self.block_size;
-        for (b, (owned, chunk)) in buffer.blocks().enumerate() {
-            if owned {
-                let block = start_block + b;
-                self.owned[block] = 1;
-                self.block_mut(block).copy_from_slice(chunk);
+
+        // Special case: if we're reading the entire buffer and the incoming
+        // buffer is fully owned, then swap pointers instead of copying data.
+        if start_block == 0
+            && self.len() == buffer.len()
+            && buffer.owned.iter().all(|o| *o != 0)
+        {
+            std::mem::swap(self, buffer);
+        } else {
+            for (b, (owned, chunk)) in buffer.blocks().enumerate() {
+                if owned {
+                    let block = start_block + b;
+                    self.owned[block] = 1;
+                    self.block_mut(block).copy_from_slice(chunk);
+                }
             }
         }
-
-        buffer.reset(0, self.block_size);
```

Comment:
The function is `pub(crate)`, so it can't be used by anyone outside of the `crucible` crate.  Within the crate, it's only used in `Volume::read` (the target of this optimization!) and a few unit tests, which I audited by hand.